Fix to Issue #622, with additional tests

davidcok opened the pull request:
Fix to Issue #622, for Java. Generates output class name by replacing all non-Java-identifier characters with '_' (not just '-' as stated in the issue). Note that this could cause class and file name collisions if a user named files for example 'a-b.dfy' and 'a#b.dfy'

RustanLeino commented on Test/naming-z/README:
```diff
@@ -0,0 +1 @@
+## This is a test for issue #622
```
Thanks for guarding the fix against future changes by adding these tests. However, rather than creating a new test folder (especially one that will hold just one test), please move this test into the Test/git-issues folder and somehow add `622` to the filename `d-e$f.dfy`.

RustanLeino commented on Source/Dafny/Compiler-java.cs:
```diff
@@ -3757,5 +3757,12 @@ protected override BlockTargetWriter CreateIterator(IteratorDecl iter, TargetWri
+    // Transforms a legal file name (without extension or directcory) into 
+    // a legal class name in the target language
+    public override string TransformToClassName(string baseName) {
+      //return baseName.Replace('-', '_');
```
Remove this line

RustanLeino commented on Source/Dafny/Compiler-java.cs:
```diff
@@ -3757,5 +3757,12 @@ protected override BlockTargetWriter CreateIterator(IteratorDecl iter, TargetWri
+    // a legal class name in the target language
+    public override string TransformToClassName(string baseName) {
+      //return baseName.Replace('-', '_');
+      return System.Text.RegularExpressions.Regex.Replace(baseName, "[^_A-Za-z0-9\\$]", "_");
```
Does this replace one occurrence or all occurrences? (I presume the latter, since the test file below has both a `-` and a `$` in the name.)

RustanLeino commented:
If you add "Fixes #622" on a line by itself in the PR/commit message, then github will automatically close that Issue when the PR is merged.

davidcok commented:
Tests renamed and embellished -- which worked because git-issues also contains a '-', as needed for the test

RustanLeino commented on Test/git-issues/git-issue-622$f.dfy:
```diff
@@ -0,0 +1,11 @@
+// RUN: rm -f 'git-issue-622$f/git_issue_622$f.java'
```
This line is not necessary, since L2 will overwrite the file anyhow. This also means we avoid a dependency on `rm`, which may not be available on Windows systems.

RustanLeino commented on Test/git-issues/git-issue-622$f.dfy:
```diff
@@ -0,0 +1,11 @@
+// RUN: rm -f 'git-issue-622$f/git_issue_622$f.java'
+// RUN: %dafny /compile:3 /compileTarget:java /spillTargetCode:3 "%s" > "%t"
+// RUN: file 'git-issue-622$f/git_issue_622$f.java' >> "%t"
```
`file` is not available on Windows. Is there some other way to achieve this? I guess the point is to check that the file created has this exact name.

RustanLeino replied:
`file` is available on the build machine, but produces slightly different output. Unless there's another easy solution, I think it's fine to just omit this part of the test. We would still be checking that it doesn't crash.

robin-aws replied:
I'm in the middle of refactoring the tests to use xUnit to address #632, and that will mean we can use .NET runtime/standard library utilities for these kinds of operations to stay platform independent. I need perhaps another day or two before the PR is ready though.

davidcok commented:
Fixed Main. Removed the rm. Also removed the file command to make the test more platform-independent, though this makes the test less useful because Dafny runs the java program internally so the file name translation is not actually tested.

davidcok commented:
I'd rather not wait for Robin's test refactoring, as I have subsequent issue fixes that depend on this one.

robin-aws commented:
> I'd rather not wait for Robin's test refactoring, as I have subsequent issue fixes that depend on this one.

No worries, I'll deal with refactoring whatever you get working into the new format as needed.

davidcok commented:
Robin, In particular the need is this:

It should be possible to run a test of Dafny that intentionally emits an error message and a non-zero exit code and check for the accuracy of both of those.

I cannot find a way to that properly in lit, but I know that cat least jUnit tests can do that, so I suppose the xUnit framework does as well

- David

> On Jun 3, 2020, at 7:55 PM, Robin Salkeld <notifications@github.com> wrote:
> 
> 
> I'd rather not wait for Robin's test refactoring, as I have subsequent issue fixes that depend on this one.
> 
> No worries, I'll deal with refactoring whatever you get working into the new format as needed.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/dafny-lang/dafny/pull/640#issuecomment-638520427>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABITDQCDZZYHMIUSKBBLM6LRU3PFPANCNFSM4NQJPPDA>.
>

robin-aws commented:
Ah I had a similar problem with testing here, would this fix it? 

https://github.com/dafny-lang/dafny/blob/master/Test/DafnyTests/TestAttribute.dfy#L2
